### PR TITLE
REC-55: refactor release workflow

### PR DESCRIPTION
**Pull request by jayconrod**
Previously, it was one job that ran on Linux with CI runners.
We need to sign and notarize macOS binaries though, and we can
only do that on a macOS machine.

So this change splits the workflow into multiple jobs. The first
job checks the version string, next we build binaries (signing
not done yet), last we create the release.

### CODE REVIEW

**Comment by jayconrod** [comment]
Test run: https://github.com/EngFlow/auth/actions/runs/11410437468

**Review by minor-fixes — APPROVED** [review]
w.r.t. the libc version requirement - do you think we should be testing for that on release? If so, it sounds like adding an `sh_test` that takes the built binary as data might not be the best path, if it depends on which remote runner it runs on - so would another job in this workflow make sense here?

File: infra/release-build.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  echo >&2 "RELEASE_VERSION not set"
+  exit 1
+fi
+
+APPLE_CERT_ID=
+install_cert () {
+  if [[ "${OS}" != 'macos' ]]; then
+    return
+  fi
+  # Files in $RUNNER_TEMP are automatically removed on completion.
```
[the remark this answers is not in the document]
Comment (jayconrod) [reply infra/release-build.sh:38]: Yes, that would work. I just learned about `RUNNER_TEMP` from another script.

For `gh` specifically though, I think we should throw it into the CI runner container image. Now that we're using CI runners for Linux, it's much easier to add build-time dependencies like that.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -30,27 +30,146 @@ permissions:
...
+  build-linux-artifacts:
+    # Linux is separate because we execute with CI runners, which don't support
+    # macOS or Windows yet. We use a Debian 11 image because binaries built on
+    # newer versions are incompatible with older version due to the libc runtime
+    # dependency.
+    runs-on:
+      - self-hosted
+      - os=linux
+      - arch=x64
+      - "engflow-container-image=docker://645088952840.dkr.ecr.eu-west-1.amazonaws.com/engflow-ci/debian12-dind-x64@sha256:763903935682de148b4e09fe1d7ef3bbc4ec829d59c3f41cb9519984639eaa06"
```
[the remark this answers is not in the document]
Comment (jayconrod) [reply .github/workflows/release.yml]: Yes! Glad you caught this.

**Comment by jayconrod** [comment]
> w.r.t. the libc version requirement - do you think we should be testing for that on release? If so, it sounds like adding an sh_test that takes the built binary as data might not be the best path, if it depends on which remote runner it runs on - so would another job in this workflow make sense here?

I thought about this a bit. It seems valuable to confirm that the binary can run on Debian 11. We could try running it from a script in the same job, but that only proves the binary runs on the same platform where it was built, which doesn't add much value. We could try running the script in a different job, but that also feels tautological: it would be just as easy to typo the container image for that job as the build job.

We could try inspecting the binary itself. The relevant output is:

```
$ objdump -T engflow_auth
...
0000000000000000      DF *UND*  0000000000000000 (GLIBC_2.34) pthread_attr_getstacksize
0000000000000000      DF *UND*  0000000000000000 (GLIBC_2.34) pthread_setspecific
0000000000000000      DF *UND*  0000000000000000 (GLIBC_2.2.5) freeaddrinfo
0000000000000000      DO *UND*  0000000000000000 (GLIBC_2.2.5) stderr
0000000000000000      DF *UND*  0000000000000000 (GLIBC_2.2.5) seteuid
```

Each glibc symbol has the minimum version needed. 2.34 is the version we don't want. But this seems overly brittle and implementation specific. I also really don't want to parse and compare semver in bash.

**Pull request merged**